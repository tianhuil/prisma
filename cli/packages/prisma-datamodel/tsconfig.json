{
  "compilerOptions": {
    "forceConsistentCasingInFileNames": true,
    "strictNullChecks": true,
    "sourceMap": true,
    "outDir": "dist",
    "moduleResolution": "node",
    "rootDir": "src",
    "lib": ["es2017", "dom", "esnext.asynciterable"],
    "target": "es5"
  },
  "exclude": ["node_modules", "dist", "__tests__"]
}
