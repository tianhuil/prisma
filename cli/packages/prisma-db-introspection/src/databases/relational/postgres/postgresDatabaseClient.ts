import IDatabaseClient from '../../IDatabaseClient'
import { Client } from 'pg'

export default class PostgresDatabaseClient implements IDatabaseClient {
  private client: Client

  constructor(client: Client) {
    this.client = client
  }

  public async query(query: string, variables: any[]): Promise<any[]> {
    return (await this.client.query(query, variables)).rows
  }
}
