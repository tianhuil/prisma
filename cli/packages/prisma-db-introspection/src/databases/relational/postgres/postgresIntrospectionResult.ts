import {
  ITable,
  ITableRelation,
  IEnum,
  ISequenceInfo,
} from '../relationalConnector'
import { RelationalIntrospectionResult } from '../relationalIntrospectionResult'
import {
  IGQLField,
  IGQLType,
  DatabaseType,
  Renderer,
  TypeIdentifiers,
  GQLAssert,
  IdStrategy,
} from 'prisma-datamodel'

export class PostgresIntrospectionResult extends RelationalIntrospectionResult {
  constructor(
    model: ITable[],
    relations: ITableRelation[],
    enums: IEnum[],
    sequences: ISequenceInfo[],
    renderer?: Renderer,
  ) {
    super(model, relations, enums, sequences, DatabaseType.postgres, renderer)
  }

  protected isTypeReserved(type: IGQLType): boolean {
    return type.name == '_RelayId'
  }
  protected toTypeIdentifyer(
    fieldTypeName: string,
    fieldInfo: IGQLField,
    typeName: string,
  ): string | null {
    switch (fieldTypeName) {
      case 'int1':
      case 'int2':
      case 'int4':
      case '_int4':
      case 'int8':
        return TypeIdentifiers.integer
      case 'numeric':
      case 'float4':
      case 'float8':
        return TypeIdentifiers.float
      case 'varchar':
      case 'bpchar':
      case '_text':
      // If we have a text type on an ID field, we map to the ID type.
      case 'text':
        return fieldInfo.isId ? TypeIdentifiers.id : TypeIdentifiers.string
      case 'bool':
        return TypeIdentifiers.boolean
      case 'jsonb':
      case 'json':
        return TypeIdentifiers.json
      case '_date':
      case 'date':
      case 'timestamptz':
      case 'timestamp':
        return TypeIdentifiers.dateTime
      case 'uuid':
        return TypeIdentifiers.uuid
      default:
        return null
    }
  }
  protected parseDefaultValue(
    defaultValueString: string,
    type: string,
  ): string | null {
    let val = defaultValueString

    // Remove cast operator
    const i = val.indexOf('::')

    if (i >= 0) {
      val = val.substring(0, i)
    }

    // Detect string
    if (val.startsWith("'")) {
      // Strip quotes (are added again by renderer)
      val = val.substring(1, val.length - 1)
    }

    // Check for null
    if (val.toUpperCase() === 'NULL') {
      return null
    }

    // If the field is not a string field,
    // and the default val is not a boolean or a number, we assume a function call or sequence reference.
    if (type !== TypeIdentifiers.string && type != TypeIdentifiers.id) {
      if (
        isNaN(val as any) &&
        val.toLowerCase() !== 'true' &&
        val.toLowerCase() !== 'false'
      ) {
        return null
      }
    }

    // TODO: Sequences are simply ignored.

    return val
  }

  protected resolveSequences(types: IGQLType[], sequences: ISequenceInfo[]) {
    for (const type of types) {
      for (const field of type.fields) {
        if (
          typeof field.defaultValue === 'string' &&
          field.defaultValue.startsWith('nextval')
        ) {
          // Regex also truncates the database schema name, if included in the regex. that's the first capture group.
          const match = field.defaultValue.match(
            /^nextval\('(?:.*\.)?(.*?)'::regclass\)$/i,
          )

          if (match === null) {
            continue
          }
          let [dummy, seqName] = match

          // Trim quotes.
          if(seqName.startsWith('"')) {
            seqName = seqName.substring(1, seqName.length - 1)
          }

          const seq = sequences.find(x => x.name === seqName)

          if (seq === undefined) {
            field.comments.push({
              text: `Error resolving sequence ${seqName} for ${type.name}.${
                field.name
              }: The sequence was not found.`,
              isError: true,
            })
          } else {
            field.idStrategy = IdStrategy.Sequence
            field.associatedSequence = seq!
            field.defaultValue = null
          }
        }
      }
    }
    return types
  }
}
