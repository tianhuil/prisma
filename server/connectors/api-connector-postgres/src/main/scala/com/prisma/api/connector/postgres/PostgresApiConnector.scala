package com.prisma.api.connector.postgres

import java.sql.Driver

import com.prisma.api.connector.ApiConnector
import com.prisma.api.connector.jdbc.impl.{JdbcDataResolver, JdbcDatabaseMutactionExecutor}
import com.prisma.config.DatabaseConfig
import com.prisma.shared.models.{ConnectorCapabilities, Project, ProjectIdEncoder}

import scala.concurrent.{ExecutionContext, Future}

case class PostgresApiConnector(config: DatabaseConfig, driver: Driver)(implicit ec: ExecutionContext) extends ApiConnector {
  lazy val databases = PostgresDatabasesFactory.initialize(config, driver)

  override def initialize() = {
    databases
    Future.unit
  }

  override def shutdown() = {
    for {
      _ <- databases.primary.database.shutdown
      _ <- databases.replica.database.shutdown
    } yield ()
  }

  override val databaseMutactionExecutor            = JdbcDatabaseMutactionExecutor(databases.primary)
  override def dataResolver(project: Project)       = JdbcDataResolver(project, databases.primary)
  override def masterDataResolver(project: Project) = JdbcDataResolver(project, databases.primary)
  override def projectIdEncoder: ProjectIdEncoder   = ProjectIdEncoder('$')
  override val capabilities                         = ConnectorCapabilities.postgresPrototype
}
