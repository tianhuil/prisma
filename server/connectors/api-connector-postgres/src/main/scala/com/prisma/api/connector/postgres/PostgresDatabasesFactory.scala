package com.prisma.api.connector.postgres

import java.sql.Driver

import com.typesafe.config.ConfigFactory
import com.prisma.config.DatabaseConfig
import com.prisma.connector.shared.jdbc.{Databases, SlickDatabase}
import com.zaxxer.hikari.{HikariConfig, HikariDataSource}
import slick.jdbc.{DataSourceJdbcDataSource, DriverDataSource, PostgresProfile}
import slick.jdbc.PostgresProfile.api._
import slick.jdbc.hikaricp.HikariCPJdbcDataSource

object PostgresDatabasesFactory {
  // PostgreSQL db used for all Prisma schemas (must be in sync with the deploy connector)
  val defaultDatabase = "prisma"

  // Schema to use in the database
  val schema = "public" // default schema

  def initialize(dbConfig: DatabaseConfig, driver: Driver): Databases = {
    val masterDb                     = databaseForConfig(dbConfig, driver)
    val slickDatabase: SlickDatabase = SlickDatabase(PostgresProfile, masterDb)

    Databases(primary = slickDatabase, replica = slickDatabase)
  }

  private def isPooled(dbConfig: DatabaseConfig) = {
    dbConfig.pooled
  }

  def databaseForConfig(dbConfig: DatabaseConfig, driver: Driver) = {
    if (isPooled(dbConfig))
      pooledDataSource(dbConfig, driver)
    else
      simpleDataSource(dbConfig, driver)
  }

  def simpleDataSource(dbConfig: DatabaseConfig, driver: Driver) = {
    val database = dbConfig.database.getOrElse(defaultDatabase)

    val source = new DataSourceJdbcDataSource(
      new DriverDataSource(
        url =
          s"jdbc:postgresql://${dbConfig.host}:${dbConfig.port}/$database?currentSchema=$schema&ssl=${dbConfig.ssl}&sslfactory=org.postgresql.ssl.NonValidatingFactory",
        user = dbConfig.user,
        password = dbConfig.password.getOrElse(""),
        driverObject = driver
      ),
      keepAliveConnection = true,
      maxConnections = None
    )

    val poolName       = "database"
    val numThreads     = dbConfig.connectionLimit.getOrElse(10) - 1
    val maxConnections = numThreads
    val executor = AsyncExecutor(
      name = poolName,
      minThreads = numThreads,
      maxThreads = numThreads,
      queueSize = dbConfig.queueSizeLimitOrDefault,
      maxConnections = maxConnections,
      registerMbeans = false
    )

    Database.forSource(source, executor)
  }

  def pooledDataSource(dbConfig: DatabaseConfig, driver: Driver) = {
    val pooled   = if (dbConfig.pooled) "" else "connectionPool = disabled"
    val database = dbConfig.database.getOrElse(defaultDatabase)

    val config = ConfigFactory
      .parseString(s"""
                      |database {
                      |  url = "jdbc:postgresql://${dbConfig.host}:${dbConfig.port}/$database?currentSchema=$schema&ssl=${dbConfig.ssl}&sslfactory=org.postgresql.ssl.NonValidatingFactory"
                      |  properties {
                      |    user = "${dbConfig.user}"
                      |    password = "${dbConfig.password.getOrElse("")}"
                      |  }
                      |  numThreads = ${dbConfig.connectionLimit.getOrElse(10) - 1} // we subtract 1 because one connection is consumed already by deploy
                      |  queueSize = ${dbConfig.queueSizeLimitOrDefault}
                      |  connectionTimeout = 5000
                      |  $pooled
                      |}
      """.stripMargin)
      .resolve

    Database.forConfig("database", config, driver)
  }
}
