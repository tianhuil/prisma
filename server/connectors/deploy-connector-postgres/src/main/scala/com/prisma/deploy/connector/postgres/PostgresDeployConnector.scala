package com.prisma.deploy.connector.postgres

import java.sql.Driver

import com.prisma.config.DatabaseConfig
import com.prisma.deploy.connector._
import com.prisma.deploy.connector.jdbc.PostgresDatabaseInspector
import com.prisma.deploy.connector.jdbc.database.{JdbcClientDbQueries, JdbcDeployMutactionExecutor}
import com.prisma.deploy.connector.jdbc.persistence.{JdbcCloudSecretPersistence, JdbcMigrationPersistence, JdbcProjectPersistence, JdbcTelemetryPersistence}
import com.prisma.deploy.connector.persistence.{CloudSecretPersistence, MigrationPersistence, ProjectPersistence, TelemetryPersistence}
import com.prisma.deploy.connector.postgres.database._
import com.prisma.shared.models.{ConnectorCapabilities, Project, ProjectIdEncoder}
import org.joda.time.DateTime
import slick.dbio.Effect.Read
import slick.dbio.{DBIOAction, NoStream}
import slick.jdbc.PostgresProfile.api._

import scala.concurrent.{ExecutionContext, Future}
import scala.util.{Failure, Success}

case class PostgresDeployConnector(
    dbConfig: DatabaseConfig,
    driver: Driver
)(implicit ec: ExecutionContext)
    extends DeployConnector {

  override def capabilities: ConnectorCapabilities = ConnectorCapabilities.postgresPrototype

  lazy val internalDatabases   = PostgresInternalDatabaseDefs(dbConfig, driver)
  lazy val setupDatabases      = internalDatabases.setupDatabase
  lazy val managementDatabases = internalDatabases.managementDatabase
  lazy val projectDatabases    = internalDatabases.managementDatabase

  lazy val managementDatabase = managementDatabases.primary.database
  lazy val projectDatabase    = projectDatabases.primary.database

  lazy val postgresTypeMapper = PostgresTypeMapper()
  lazy val mutationBuilder    = PostgresJdbcDeployDatabaseMutationBuilder(managementDatabases.primary, postgresTypeMapper)

  override lazy val projectPersistence: ProjectPersistence           = JdbcProjectPersistence(managementDatabases.primary, dbConfig)
  override lazy val migrationPersistence: MigrationPersistence       = JdbcMigrationPersistence(managementDatabases.primary)
  override lazy val cloudSecretPersistence: CloudSecretPersistence   = JdbcCloudSecretPersistence(managementDatabases.primary)
  override lazy val telemetryPersistence: TelemetryPersistence       = JdbcTelemetryPersistence(managementDatabases.primary)
  override lazy val deployMutactionExecutor: DeployMutactionExecutor = JdbcDeployMutactionExecutor(mutationBuilder)
  override lazy val databaseInspector: DatabaseInspector             = PostgresDatabaseInspector(projectDatabases.primary)

  override def createProjectDatabase(id: String): Future[Unit] = {
    val action = mutationBuilder.createDatabaseForProject(id = id)
    projectDatabase.run(action)
  }

  override def deleteProjectDatabase(id: String): Future[Unit] = {
    val action = mutationBuilder.deleteProjectDatabase(projectId = id).map(_ => ())
    projectDatabase.run(action)
  }

  override def getAllDatabaseSizes(): Future[Vector[DatabaseSize]] = {
    val action = {
      val query = sql"""
           SELECT schemaname, SUM(pg_total_relation_size(quote_ident(schemaname) || '.' || quote_ident(tablename))) / 1024 / 1024 FROM pg_tables GROUP BY schemaname
         """
      query.as[(String, Double)].map { tuples =>
        tuples.map { tuple =>
          DatabaseSize(tuple._1, tuple._2)
        }
      }
    }

    projectDatabase.run(action)
  }

  override def clientDBQueries(project: Project): ClientDbQueries      = JdbcClientDbQueries(project, projectDatabases.primary)
  override def getOrCreateTelemetryInfo(): Future[TelemetryInfo]       = telemetryPersistence.getOrCreateInfo()
  override def updateTelemetryInfo(lastPinged: DateTime): Future[Unit] = telemetryPersistence.updateTelemetryInfo(lastPinged)
  override def projectIdEncoder: ProjectIdEncoder                      = ProjectIdEncoder('$')

  override def initialize(): Future[Unit] = {
    // We're ignoring failures for createDatabaseAction as there is no "create if not exists" in psql
    setupDatabases.primary.database
      .run(InternalDatabaseSchema.createDatabaseAction(internalDatabases.dbName))
      .transformWith { _ =>
        val action = InternalDatabaseSchema.createSchemaActions(internalDatabases.managementSchemaName, recreate = false)
        projectDatabase.run(action)
      }
      .flatMap(_ => setupDatabases.shutdown)
  }

  override def reset(): Future[Unit] = truncateManagementTablesInDatabase(managementDatabase)

  override def shutdown() = {
    managementDatabases.shutdown
  }

  override def managementLock(): Future[Unit] = {
    managementDatabase.run(sql"SELECT pg_advisory_lock(1000);".as[String].head.withPinnedSession).transformWith {
      case Success(_)   => Future.successful(())
      case Failure(err) => Future.failed(err)
    }
  }

  protected def truncateManagementTablesInDatabase(database: Database)(implicit ec: ExecutionContext): Future[Unit] = {
    for {
      schemas <- database.run(getTables())
      _       <- database.run(dangerouslyTruncateTables(schemas))
    } yield ()
  }

  private def getTables()(implicit ec: ExecutionContext): DBIOAction[Vector[String], NoStream, Read] = {
    sql"""SELECT table_name
          FROM information_schema.tables
          WHERE table_schema = '#${internalDatabases.managementSchemaName}'
          AND table_type = 'BASE TABLE';""".as[String]
  }

  private def dangerouslyTruncateTables(tableNames: Vector[String]): DBIOAction[Unit, NoStream, Effect] = {
    DBIO.seq(tableNames.map(name => sqlu"""TRUNCATE TABLE "#$name" cascade"""): _*)
  }
}
