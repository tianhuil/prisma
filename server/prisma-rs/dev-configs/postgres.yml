port: 4466
databases:
  default:
    connector: postgres-native
    host: 127.0.0.1
    port: 5432
    user: postgres
    password: prisma
    database: prisma
    schema: CreateMutationSpec_S
    rawAccess: true
