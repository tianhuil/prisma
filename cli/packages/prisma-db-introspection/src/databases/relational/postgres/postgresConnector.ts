import {
  RelationalConnector,
  ITable,
  ITableRelation,
  IInternalEnumInfo,
  IEnum,
  ISequenceInfo,
} from '../relationalConnector'
import { Client } from 'pg'
import { DatabaseType } from 'prisma-datamodel'
import { PostgresIntrospectionResult } from './postgresIntrospectionResult'
import { RelationalIntrospectionResult } from '../relationalIntrospectionResult'
import IDatabaseClient from '../../IDatabaseClient'
import PostgresDatabaseClient from './postgresDatabaseClient'

// Documentation: https://www.prisma.io/docs/data-model-and-migrations/introspection-mapping-to-existing-db-soi1/

// Responsible for extracting a normalized representation of a PostgreSQL database (schema)
export class PostgresConnector extends RelationalConnector {
  constructor(client: IDatabaseClient | Client) {
    if (client instanceof Client) {
      client = new PostgresDatabaseClient(client)
    }
    super(client)
  }

  public getDatabaseType(): DatabaseType {
    return DatabaseType.postgres
  }

  protected createIntrospectionResult(
    models: ITable[],
    relations: ITableRelation[],
    enums: IEnum[],
    sequences: ISequenceInfo[],
  ): RelationalIntrospectionResult {
    return new PostgresIntrospectionResult(models, relations, enums, sequences)
  }

  public async listSchemas(): Promise<string[]> {
    const schemas = await super.listSchemas()
    return schemas.filter(schema => !schema.startsWith('pg_'))
  }

  protected getTypeColumnName() {
    return 'udt_name'
  }

  protected getAutoIncrementCondition() {
    return 'false'
  }

  protected parameter(count: number, type: string) {
    return `$${count}::${type}`
  }

  // TODO: Unit test for column comments
  protected async queryColumnComment(
    schemaName: string,
    tableName: string,
    columnName: string,
  ) {
    const commentQuery = `
      SELECT
      (
        SELECT
          pg_catalog.col_description(c.oid, cols.ordinal_position::int)
        FROM pg_catalog.pg_class c
        WHERE
          c.oid     = (SELECT ('"' || cols.table_schema || '"."' || cols.table_name || '"')::regclass::oid) AND
          c.relname = cols.table_name
      ) as column_comment   
      FROM
        information_schema.columns cols
      WHERE
        cols.table_schema  = $1::text AND
        cols.table_name    = $2::text AND
        cols.column_name   = $3::text;
    `
    const [comment] = (await this.query(commentQuery, [
      schemaName,
      tableName,
      columnName,
    ])).map(row => row.column_comment as string)

    if (comment === undefined) {
      return null
    } else {
      return comment
    }
  }

  protected async queryIndices(schemaName: string, tableName: string) {
    const indexQuery = `
      SELECT
          tableInfos.relname as table_name,
          indexInfos.relname as index_name,
          array_to_string(array_agg(columnInfos.attname), ',') as column_names,
          rawIndex.indisunique as is_unique,
          rawIndex.indisprimary as is_primary_key
      FROM
          -- pg_class stores infos about tables, indices etc: https://www.postgresql.org/docs/9.3/catalog-pg-class.html
          pg_class tableInfos,
          pg_class indexInfos,
          -- pg_index stores indices: https://www.postgresql.org/docs/9.3/catalog-pg-index.html
          pg_index rawIndex,
          -- pg_attribute stores infos about columns: https://www.postgresql.org/docs/9.3/catalog-pg-attribute.html
          pg_attribute columnInfos,
          -- pg_namespace stores info about the schema
          pg_namespace schemaInfo
      WHERE
          -- find table info for index
          tableInfos.oid = rawIndex.indrelid
          -- find index info
          AND indexInfos.oid = rawIndex.indexrelid
          -- find table columns
          AND columnInfos.attrelid = tableInfos.oid
          AND columnInfos.attnum = ANY(rawIndex.indkey)
          -- we only consider oridnary tables
          AND tableInfos.relkind = 'r'
          -- we only consider stuff out of one specific schema
          AND tableInfos.relnamespace = schemaInfo.oid
          AND schemaInfo.nspname = $1::text
          AND tableInfos.relname = $2::text
      GROUP BY
          tableInfos.relname,
          indexInfos.relname,
          rawIndex.indisunique,
          rawIndex.indisprimary
    `
    return (await this.query(indexQuery, [schemaName, tableName])).map(row => {
      return {
        tableName: row.table_name as string,
        name: row.index_name as string,
        fields: this.parseJoinedArray(row.column_names),
        unique: row.is_unique as boolean,
        isPrimaryKey: row.is_primary_key as boolean,
      }
    })
  }

  private parseJoinedArray(arrayAsString: string): string[] {
    if (arrayAsString === null || arrayAsString === undefined) {
      return []
    }
    return arrayAsString.split(',').map(x => x.trim())
  }

  protected async queryEnums(schemaName: string): Promise<IInternalEnumInfo[]> {
    const enumQuery = `
      SELECT
        t.typname AS "enumName",  
        array_to_string(array_agg(e.enumlabel), ',') AS "enumValues"
      FROM pg_type t 
        JOIN pg_enum e ON t.oid = e.enumtypid  
        JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
      WHERE 
          n.nspname = $1::text
      GROUP BY t.typname`

    return (await this.query(enumQuery, [schemaName])).map(row => {
      return {
        name: row.enumName as string,
        values: this.parseJoinedArray(row.enumValues),
      }
    })
  }

  protected async listSequences(schemaName: string): Promise<ISequenceInfo[]> {
    const sequenceQuery = `
    SELECT
      sequence_name, start_value
      FROM 
      information_schema.sequences
      WHERE
      sequence_schema =  $1::text`

    return (await this.query(sequenceQuery, [schemaName])).map(row => {
      return {
        name: row.sequence_name as string,
        initialValue: row.start_value as number,
        allocationSize: 1,
      }
    })
  }
}
