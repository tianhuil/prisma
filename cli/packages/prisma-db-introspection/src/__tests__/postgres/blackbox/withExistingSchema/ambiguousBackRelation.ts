import Connectors from '../../../../connectors'
import { Client } from 'pg'
import { connectionDetails } from '../connectionDetails'
import { PostgresConnector } from '../../../../databases/relational/postgres/postgresConnector'
import { DatabaseType, DefaultParser } from 'prisma-datamodel'
import { connect } from 'tls'

const existingSchema = `
type User @db(name: "users") {
    id: ID! @unique
    asked: [Question] @relation(name: "question_asked")
    answered: [Question] @relation(name: "question_answer")
}
  
type Question @db(name: "questions") {
    id: ID! @unique
    asker: User! @relation(name: "question_asked") @db(name: "asker_id")
    answerer: User @relation(name: "question_answer") @db(name: "answerer_id")
}
`

async function introspect(client: Client) {
  const existing = DefaultParser.create(
    DatabaseType.postgres,
  ).parseFromSchemaString(existingSchema)
  return (await Connectors.create(DatabaseType.postgres, client).introspect(
    'test',
  )).renderToNormalizedDatamodelString(existing)
}

async function testSchema(sql: string) {
  const client = new Client(connectionDetails)
  await client.connect()
  await client.query('DROP SCHEMA IF EXISTS "test" cascade;')
  await client.query(sql)

  expect(await introspect(client)).toMatchSnapshot()

  await client.end()
}

describe('Introspector', () => {
  test('Type with scalar lists and existing schema', async () => {
    await testSchema(`     
        CREATE SCHEMA "test";
            
        CREATE TABLE "test".questions (
            id integer NOT NULL,
            asker_id integer NOT NULL,
            answerer_id integer
        );

        CREATE TABLE "test".users (
            id integer NOT NULL
        );

        ALTER TABLE ONLY "test".questions
        ADD CONSTRAINT questions_pk PRIMARY KEY (id);

        ALTER TABLE ONLY "test".users
        ADD CONSTRAINT users_pk PRIMARY KEY (id);
        
        ALTER TABLE ONLY "test".questions
        ADD CONSTRAINT questions_fk1 FOREIGN KEY (asker_id) REFERENCES users(id);

        ALTER TABLE ONLY "test".questions
        ADD CONSTRAINT questions_fk2 FOREIGN KEY (answerer_id) REFERENCES users(id);
    `)
  })
})
