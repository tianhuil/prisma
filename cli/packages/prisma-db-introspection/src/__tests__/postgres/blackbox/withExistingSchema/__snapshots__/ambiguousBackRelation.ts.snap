// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspector Type with scalar lists and existing schema 1`] = `
"type User @pgTable(name: \\"users\\") @relationTable {
  id: Int! @unique
  asked: [Question] @relation(name: \\"question_asked\\")
  answered: [Question] @relation(name: \\"question_answer\\")
}

type Question @pgTable(name: \\"questions\\") @relationTable {
  id: Int! @unique
  asker: User! @pgColumn(name: \\"asker_id\\") @relation(name: \\"question_asked\\")
  answerer: User @pgColumn(name: \\"answerer_id\\") @relation(name: \\"question_answer\\")
}"
`;
